Paired read tracks

cmdcolin opened the pull request:
I wanted to initiate a pull request for the paired read viewing functionality cause it's kinda exciting :)

This tries to address the long standing need to view paired end reads as connected entities

See #521 for the basic idea

The code in @gmod/bam features a "read dispatching method" to grab connected pairs. The functionality from that module is compared against some output from "bazam" https://github.com/ssadedin/bazam/


Updated:

Possible TODO items

- If the pairs of the reads overlap (e.g. short insert size), then it is a little confusing
- Reads in the view as read cloud are not clickable. It could be done not too hard but even so they all overlap each other

DONE

- ~Use RadioMenuItems in track menu~
- ~Auto infer view as pairs from glyph type~
- ~Make it work for CRAM~
- ~If a read is split on two sides of a block in the block based track, then the block in the middle won't get info about that read so the feature will not display properly between them~

cmdcolin commented:
Here is a data directory that can be used to test large insert sizes

[large_inserts.tar.gz](https://github.com/GMOD/jbrowse/files/2497376/large_inserts.tar.gz)

The requests for paired end reads is highly reliant on what is in the "current view". It appears read pairs are resolved well, the efficiency is possibly a little slower than what I want, and especially on a full sized dataset? For some reason this smaller sample data is better. But functionally it is working pretty well.

![screenshot-localhost-2018 10 19-17-27-53](https://user-images.githubusercontent.com/6511937/47244671-57822d80-d3c4-11e8-8224-129332343bff.png)

cmdcolin commented:
I haven't evaluated this possibility but a "push it to the limit" extension of this PR is to render "linked reads" e.g. 10x genomics data. This has not just "paired reads" but many reads with the same molecular identifier

![10x22](https://user-images.githubusercontent.com/6511937/47319513-df627480-d61c-11e8-86a7-2644f0d6b606.jpg)
example screenshot from https://www.biocompare.com/Bench-Tips/341771-Optimizing-your-Linked-Read-Genome-and-Exome-Analyses-4-Practical-Considerations/

Similar things can also happen with long reads too with linked supplementary alignments https://www.pacb.com/blog/igv-3-improves-support-pacbio-long-reads/

cmdcolin commented:
Random other thing: color alignments by tag can reveal some interesting information about haplotypes if the tags exist https://whatshap.readthedocs.io/en/latest/guide.html

cmdcolin commented:
I did a couple other things for this PR including

- Adding triangulated ends to indicate strand orientation to the alignments...no more trying to think if red or blue means positive/negative
- Added a menu item to toggle SNPCoverage which is related to #1154

![screenshot-localhost-2018 10 24-14-31-21](https://user-images.githubusercontent.com/6511937/47453008-97b52780-d799-11e8-9f42-3f2aa03d6dd2.png)

cmdcolin commented:
The aforementioned issue of resolving "If a read is split on two sides of a block" has a solution now by keeping track of things in a featureCache at the store level, and filling in features from the cache that overlap block requests. It also clears out the featureCache once things stop overlapping the blocks. Testing on the large_inserts.tar.gz data is the best way to see this aspect of the code

cmdcolin commented:
@rbuels if you have any input on CRAM support. The basic idea of the retrofitting that was added to the BAM adapters was to fetch reads in a region and then fetch all their connecting mate pairs when a getRecordsForRange has a viewAsPairs flag enabled. I know CRAM implementation of this might be two-faceted related to the fact that reads within segments can be resolved automatically

rbuels commented:
these menu items need to be disabled for all the stores that don't support `viewAsPairs`.

should make an attribute of all stores that indicates whether view as pairs is supported, and have the track check that

rbuels commented:
probably should also rename "scaleFactor" to something more descriptive too

rbuels commented:
probably worth implementing a dialog box and an auto-estimation for `scaleFactor` as well

cmdcolin commented on src/JBrowse/Store/SeqFeature/BAM.js:
```diff
@@ -117,8 +145,8 @@ return declare( [ SeqFeatureStore, DeferredStatsMixin, DeferredFeaturesMixin, In
                 csiFilehandle: csiBlob,
                 renameRefSeqs: n => this.browser.regularizeReferenceName(n),
-                fetchSizeLimit: 100000000 || args.fetchSizeLimit,
-                chunkSizeLimit: 10000000 || args.chunkSizeLimit
```
100 megabytes fetch limit...20 mb chunk limit...big data !

cmdcolin commented on src/JBrowse/Store/SeqFeature/BAM.js:
```diff
@@ -7,8 +7,24 @@ const bamIndexedFilesCache = LRU(5)
+            return Math.min(this.f1.get('start'), this.f2.get('start'))
+        } else if(field == 'end') {
+            return Math.max(this.f1.get('end'), this.f2.get('end'))
+        } else if(field == 'name') {
```
good catch

cmdcolin commented:
During dev meeting I mentioned that very large data requests  were being launched at startup (e.g. 100MB) and this was due to the funkyness that is related to #1187. It may be of interest to consider merging this PR in order to mitigate this issue, because without #1187 there are just a bunch of requests to eventually unused regions of data. It does point to the fact that sometimes large region requests incur unreasonably large downloads on the paired reads which might need mitigation in other ways, but it is just especially noticeable due to #1187

cmdcolin commented:
Added the essentials for CRAM support on this branch!

cmdcolin commented:
TODO type things

Large insert sizes can cause visual glitches especially when you are scrolling and then you encounter a feature that is suddenly paired on the other side of where you were, so then the old blocks that rendered nothing don't get updated to fill in. 

Some approaches
* Filter larger insert sizes, and then this basically will not happen (even at high zoom levels it shouldn't happen because it requests large enough regions)
* Make some mechanism to re-render all blocks if this situation occurs
* Make some other glyph that combo's with the normal glyphs that indicates the crazy pairing e.g. a vertical line or arc

Another possible TODO is the optimization to not resolve pairs of reads from getFeatures for ultra large inserts as this generally just requests a lot of data and then doesn't actually seem necessary. Instead just create a feature span that uses PNEXT (position of pair read) instead of the actual PNEXT resolved read

rbuels commented:
We need checkboxes or radio buttons in the "Track visualization types" menu to see what visualization you are currently using:
![image](https://user-images.githubusercontent.com/73749/48641354-537b0880-e98e-11e8-81a0-afb0990d2da0.png)

rbuels commented:
Also, need word wrap in the detail dialogs

rbuels commented:
Here's a screenshot of the arc bug:
![image](https://user-images.githubusercontent.com/73749/48641581-19f6cd00-e98f-11e8-901b-6c17accd08c3.png)

rbuels commented:
Also, we need to store the user's display choice in localstorage.

rbuels commented:
The arc bug seems to be chrome-specific

garrettjstevens commented:
I think something weird might be going on here (or maybe it's a weird CRAM?). Config is:
```
[GENERAL]
refSeqs=https://s3.amazonaws.com/1000genomes/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa.fai
[tracks.illumina_LC_CRAM]
urlTemplate=https://s3.amazonaws.com/1000genomes/phase3/data/NA18504/alignment/NA18504.mapped.ILLUMINA.bwa.YRI.low_coverage.20120522.bam.cram
```
and the location is `chr1:75351739..75352025`.

Viewing unpaired looks like this:
![image](https://user-images.githubusercontent.com/25592344/48653351-d79ab380-e9c1-11e8-8608-8ded5f121e0e.png)
And viewing paired looks like this:
![image](https://user-images.githubusercontent.com/25592344/48653374-f731dc00-e9c1-11e8-9f41-a2ef2ab43aa7.png)

Gray means it's a normally paired read, right? I click on a bunch of them and they all seem to show a next segment position that's pretty close, but they disappear in paired view. Also, this looked odd:
![tmp](https://user-images.githubusercontent.com/25592344/48653561-63f9a600-e9c3-11e8-9562-aa3eb75e18cd.png)

Shouldn't those be the same? Unless it's some multi-read segment I guess. That's from one of the gray reads in the paired reads view.

garrettjstevens commented:
It looks like some of the non-wrapping stuff got fixed, but BQ still isn't wrapping:
![image](https://user-images.githubusercontent.com/25592344/48653617-bcc93e80-e9c3-11e8-9d91-48ac598ba6d4.png)

nathanhaigh commented:
Just been looking at the PR with my wheat data. I like it.

http://crobiad.agwine.adelaide.edu.au/dawn/jbrowse-test/?loc=chr2D_part2%3A17914799..17922662&tracks=DNA%2CIWGSC_v1.0_HC_genes%2CWyalkatchem_reads%2CIWGSC_grain_Z71
I have a couple of questions/comments:

- Shouldn't the "Colour by" options be radio buttons - I assume you can't overlay/select multiple options at once and that this makes sense?
- "RNA-Seq Orientation" may be better worded as "Stranded RNA-Seq"?
- How does "colour by read orientation and insert size" work with a BAM containing multiple insert sized libraries? You can see at least 2 distinct insert sizes in the "read cloud" mode below(the above link takes you to this rough area - probably need to decrease the `featureScale` value). It probably makes sense to do this per RG rather than just the whole file?
![image](https://user-images.githubusercontent.com/2037575/48655334-60b6f680-ea65-11e8-9b28-ae29784df3b2.png)
- "Show interchromosomal" may be better reworded as the reference sequences don't actually have to be chromosomes. Perhaps "Show intersequence read alignments"?

cmdcolin commented:
@garrettjstevens for the first issue, with the difference between actual feature location and the next pos (PNEXT/RNEXT),  I'm fairly certain we're just reporting the data as-is. It's hard to say. TLEN is also something recorded weirdly and for CRAM, TLEN is a combination of values depending on whether it is inter or intra slice.

For the second where they are not getting paired, it looks like these are lacking read name for whatever reason, and read name is important for pairing....weird! Might need to look at the cram code?

cmdcolin commented:
@nathanhaigh Thanks for testing this! For the questions/feedback

-I think the color by should be radio options too, I just made the 'track visualization options' into radio buttons also
-Stranded RNA-seq refers to a specific protocol I think, the XS and TS tags can be due to inferring orientation based on the canonical splice site ordering in the mapping so maybe orientation makes sense to keep
-Insert size estimation if there are multiple insert size read groups is interesting as a problem. It might be in some sense easier to break the bam into multiple according to read group but if it is very helpful to have multiple read group in a single file I could think about this. Some mate pair libraries are actually tricky to deal with because they contain paired end contamination so they implicitely have multiple insert size distributions
- Show interchromosomal is sort of just an option I added ad-hoc for the paired arc view since it just draws a orange vertical line for interchromosomal connections. Since it doesn't apply to other views, and since the vertical line is already kind of intrusive, it might be worth figuring out even a different method so this might change

cmdcolin commented:
@garrettjstevens Also as far as BQ not wrapping can you test again? I changed the code to try only long strings by dynamically adding a css class "long"

nathanhaigh commented:
Thanks for the reply @cmdcolin The only thing I'd like to tease apart further is the insert-size estimation:

> Insert size estimation if there are multiple insert size read groups is interesting as a problem. It might be in some sense easier to break the bam into multiple according to read group but if it is very helpful to have multiple read group in a single file I could think about this. Some mate pair libraries are actually tricky to deal with because they contain paired end contamination so they implicitely have multiple insert size distributions.

It makes sense to me to aggregate multiple sequencing lanes, library preps (which will inherently have independent insert-size distributions) into a single BAM/track to represent a single accession/genotype. Especially, since getting sufficient coverage for wheat WGS often requires several lanes of data, each of which get their own read group `ID`.

While mate pair libraries will contain PE read contamination, it is only possible to see these by also considering their relative orientation in addition to the insert-size estimation. So I would argue that they do not have "different" insert-size distributions per se, but that they are feature of the distribution where their particular insert-size is smaller than most of the other read "pairs" AND their orientation is `FR` rather than `RF`. As such, I would be calculating the mean/stddev across ALL reads, and not to attempt to bring orientation into this calculation. Typically, the `FR` oriented reads should be in the minority, if the MP library prep went well, so wouldn't significantly affect the insert-size estimates. If the library prep didn't go so well, then I'd argue that it's up to the researcher to process those libraries more appropriately since their QC metrics would be poorer than expected.

Strictly speaking, the insert-size estimation should be done per `LB` (library) if it is present in the `RG` header of a BAM file, otherwise falling back to `RG ID` makes sense.

cmdcolin commented:
>> Typically, the FR oriented reads should be in the minority, if the MP library prep went well, so wouldn't significantly affect the insert-size estimates.

I have seen it where it was quite a lot of small insert PE even being a majority in some regions. However I basically came to same conclusion to not factor the orientationType into insert size stats estimation.  I did make a "minimum insert size" to avoid biasing calculations too much though

In either case, I'm just wondering what the expected coloring would be? Would you use red to color the longer inserts across all libraries but just estimate it per library? It seems like it would get overly complicated color-wise if not just using red across all libraries

cmdcolin commented:
I tried to do a couple more improvements since the last meeting

0) Make the read redispatching only redispatch as far as maxInsertSize, so that it isn't requesting blocks all over the genome. Note that the arc view is able to still exceed maxInsertSize since it uses a span concept of just using RNEXT/PNEXT fields instead of doing true read pairing
1) In order to avoid render errors, always inflate getFeatures bp range so that we can catch issue where one block finds a read pair and the neighbors don't. The Alignments2.js code has this logic as cachePromise. I also did a adjustment here that could improve the high coverage files, would be curious if this helped some crashes we were seeing.
2) Check for if read is unpaired before coloring by insert size
3) Try to implement the idea of saving track config, for as may parameters as possible including those in Edit config. This is sort of going down a sort of rabbit hole and maybe instead of my current approach the "publish" event for tracks should do cookie updates
4) Instead of filtering out large insert sizes break them apart so they are just separate unpaired alignment features that are colored red
5) Auto-infers the "viewAsPairs" and "viewAsSpans" from the current glyph. This makes for easier hand configs because previously a combination of the glyph and the viewAsPairs had to be requested.
Some possible todos
6) Don't render click maps for read cloud and pair read views for some speedup


Anyways more testing is welcome

cmdcolin commented:
Added the feature for auto-estimating the scaling factors in the read cloud views, and a toggle between linear and log views. Also added a color by mapping quality.

cmdcolin commented:
Remaining possible issues

- The ability to toggle track types and coloring schemes is a little bolted onto the side of the alignments track in particular. It could possibly use some more engineering
- The coloring schemes could use legends and docs
- Lossy read name CRAM is currently unsupported and behaves badly as pairing assumes a read name
- The track stats is only calculated once on track startup and if stats calculation doesn't receive enough features at startup it doesn't update once it might later receive enough features

Possible enhancements

- The split on read group for insert size stats mentioned by @nathanhaigh 
- Arbitrary "Color by tag" BAM feature, e.g. color by readgroup or haplotype like whatshap
- Non-paired tracks could avoid showing the paired read options. Currently they just don't do much and display the tracks as normal
- The read cloud could use an insert size y axis
- User's custom style->color callbacks are sort of an unspecified element in the "Color options" menu

Testing recommendations

- The concept of saving track config modifications in localStorage needs testing
- The memory and CPU usage on different datasets

If anyone has input or prioritization of these subtasks let me know

cmdcolin commented:
For some reason I'm seeing a CRAM track is requesting much more data than BAM in a test case I have built.

keiranmraine commented:
QQ, how does this handle overlapping reads? It's becoming more of an issue recently

cmdcolin commented:
Currently it just plots the overlapping pairs arbitrarily over one another. One read could cover up SNPs from the other, but ideally if the sequencing is correct then correct both would have the same SNP in the same position so it would not be highly consequential. I could not really figure out another idealized way to do it. Have you had a chance to test this branch yet? :)

keiranmraine commented:
This still uses Alignments2 from what I can see.  I should be able to throw up a test server today.

On the topic of overlapping reads that disagree, could the base colours blend, and have a mouse-over which shows the F/R change?  Like you say it's pretty unlikely but it does occur.

keiranmraine commented:
I can confirm it's working for us (BAM).  I have very happy people :)

AndyMenzies commented:
This makes looking at non-size selected data much easier as we can easily identify the short fragments.

Any chance of making the overlapping portion of the pair more obvious, maybe a different colour or an outline.  We can clearly see where the read drawn on top ends, but we also need to know where the read at the back ends too.  We currently can't make out the extent of the overlap on the paired view so we can't tell if SNPs/variants are likely to be in both reads.

cmdcolin commented:
I was looking into what could be done about overlapping reads. It seemed like alpha blending could be an option but I was a little disappointed with it because (a) it pretty subtle and hard to see (b) it allows the gridlines to show through over the features (c) doesn't really help with the SNPs from both being shown https://github.com/GMOD/jbrowse/tree/alpha_pairing

Example screenshot
![screenshot-localhost-2018 11 30-12-12-14](https://user-images.githubusercontent.com/6511937/49304327-06248e00-f49a-11e8-87be-353a5a90df16.png)

I really appreciate the feedback and testing though! I will keep thinking about this

rbuels commented:
Maybe some kind of pattern, like a crosshatch or something? Would need to check whether that is performant though

AndyMenzies commented:
The gridlines showing though doesn't look like an issue in the image you've provided, but I can see how they might interfere with the representation of clipped reads or insertions.  The blending in the over lap is a great starting point, but as you say its is difficult to make out whats going on.  And as Keiran said, there does need to be some representation for read overlaps where the pair don't agree at a position.

cmdcolin commented:
This overlapping read SNP display is a bit tricky....perhaps, the "mismatches" data structure from both reads could be checked against each other, and then a different little SNP glyph could be drawn....That is probably the best thing I can think of for detecting SNP in overlapping reads

cmdcolin commented:
I tried making some read overlap logic more explicit. If a read overlaps, then we take special care to draw SNPs differently. Firstly, the overlap region itself is drawn as a gray bar. Then the SNPs within the overlaps are checked against each other. The logic here is SNP focused, e.g. insertions or deletions could fool it currently, but it checks if the mismatch.start of one read's mismatches equals the mismatch.start of another read's mismatch, and if so then it says "ok just draw it as a normal SNP". Otherwise, if there is a mismatch from one read that can not be matched to the other's in the overlap region, then it draws a black box with "A/-" for example if the mismatch was simply missing in one example or matches the reference, or could be "A/T" if there are just mismatching SNP concepts

Example:

![screenshot-localhost-2018 12 04-13-48-08](https://user-images.githubusercontent.com/6511937/49465368-73525f00-f7cb-11e8-8e7c-9830d1c50897.png)
Matching SNP just shows up as normal

![screenshot-localhost-2018 12 04-13-44-49 1](https://user-images.githubusercontent.com/6511937/49465369-73525f00-f7cb-11e8-836c-db7d8d1b57a3.png)

A mismatching SNP shows up as black with some text inside

cmdcolin commented:
It probably helps at least for testing purposes to browse around with a "unpaired" track and a paired track mode on

@keiranmraine @AndyMenzies if you want to try testing that it would be awesome

keiranmraine commented:
I think the reads changing colour and black box when there is disagreement is a great idea, @AndyMenzies what do you think?  I'll deploy from that version as soon as I can, but at a conference today so won't be until at least Thursday.

AndyMenzies commented:
I agree the @keiranmraine changing the colours to be different from the standard pallet really highlights the overlaps and read discrepancies.  This will be really useful for variant chasing.

cmdcolin commented:
I bumped @gmod/cram version on this branch and it gives us the ability to use lossy read names and fixes #1271

cmdcolin commented on src/JBrowse/Store/SeqFeature/CRAM.js:
```diff
@@ -269,31 +294,55 @@ return declare( [ SeqFeatureStore, DeferredStatsMixin, DeferredFeaturesMixin, Gl
+                // map the CramSizeLimitError to JBrowse Errors.DataOverflow
+                console.error(err)
+
                 errorCallback(err)
```
Changed back. cjsRequire was disagreeing with yarn link version of @gmod/cram

cmdcolin commented:
Alrighty I think this is ready to merge after discussion today! I am going to hit the merge button and then we can move towards 1.16.0 release after some other milestone issues are closed!
